### ISSUE TITLE
[GRAALVM] Logback not fetching Environment Variable

### ISSUE THREAD

**Issue by driverpt** [micronaut-projects/micronaut-core#8139]
### Expected Behavior

We are using [Logstash Logback Encoder](https://github.com/logfellow/logstash-logback-encoder) and we are trying to inject Field Values as Environment Variables.

e.g.:
```
<pattern>
                    <omitEmptyFields>true</omitEmptyFields>
                    <pattern>
                        {
                            "CorrelationId": "%mdc{trace_id}",
                            "Resource": {
                                "application": "demo",
                                "logger": "%logger",
                                "version":"${VERSION}",
                                "environment":"${ENVIRONMENT}",
                            }
                        }
                    </pattern>
                </pattern>
```

### Actual Behaviour

Variables are not properly filled. Result:

```
{
    "Timestamp": "2022-10-10T10:01:16.565Z",
    "SeverityText": "INFO",
    "Resource": {
        "application": "demo",
        "logger": "io.micronaut.function.aws.proxy.AbstractLambdaContainerHandler",
        "version": "VERSION_IS_UNDEFINED"
        "environment": "ENVIRONMENT_IS_UNDEFINED"
    },
    "Body": {
        "message": "Starting Lambda Container Handler",
        "thread": "main"
    }
}

```

### Example Application

https://github.com/driverpt/micronaut-otel-issue

### Version

3.7.1

**Comment by driverpt** [comment]
Update:

I'm not a GraalVM specialist, but my gut tells me that the issue is here:

![image](https://user-images.githubusercontent.com/1164534/195541664-5a82dc0e-cefc-42a0-acf5-d1896377fa26.png)

Logback by being initialized in Build Time, variable substitution is also happening in this phase
Source: https://logback.qos.ch/manual/configuration.html#variableSubstitution

**Comment by n0tl3ss** [comment]
My understanding this won't be a trivial change, here is an explanation:

- GraalVM compiler will try to initialise at build time everything that is referenced in static blocks of class code that is specified inside the `--initialize-at-build-time`.
-  The `--initialize-at-build-time=ch.qos.logback,io.micronaut`, `io.micronaut` is crucial for us,  inside https://github.com/micronaut-projects/micronaut-core/blob/1d8a98ba11ec0d6d280812b9a433f2e23ba3dcca/runtime/src/main/resources/META-INF/native-image/io.micronaut/micronaut-runtime/native-image.properties#L19 tells GraalVM compiler to try as much as possible to do at build time because of optimisations and lower cpu usage in native image.
- Micronaut is coupled a lot with code that looks like `private static final Logger LOGGER = LoggerFactory.getLogger(...)` and that Logger will always try to initialise itself in build time and not in runtime even if you specify that you want to be at the run time.
-  GraalVM compiler has to choose either `-initialize-at-build-time` or `--initialize-at-run-time` for all classes that has static blocks and coupled together. It will always choose to do `-initialize-at-build-time`.

**Comment by driverpt** [comment]
@n0tl3ss , how do you suggest we fix this then?

The issue is that Logback Internal Resolvers are initialized and cached at build time, causing it to ignore the new values at runtime (given that these are already cached and resolved).

And we need to inject values at Runtime that will be used in Structured Logging

**Comment by driverpt** [comment]
Any updates on this one? @n0tl3ss  @sdelamo  @graemerocher

**Comment by graemerocher** [comment]
fixed in 4.0.x where runtime initialization is used
